src/DataProtection/DataProtection/src/KeyManagement/KeyRingProvider.cs

```diff
@@ -292,6 +312,149 @@ internal IKeyRing GetCurrentKeyRingCore(DateTime utcNow, bool forceRefresh = fal
+            if (taskKeyRing is not null)
+            {
+                return taskKeyRing;
+            }
+        }
+
+        // The cached keyring hasn't been created or must be refreshed. We'll allow one thread to
+        // create a task to update the keyring, and all threads will continue to use the existing cached
+        // keyring while the first thread performs the update. There is an exception: if there
+        // is no usable existing cached keyring, all callers must block until the keyring exists.
+        lock (_cacheableKeyRingLockObj)
+        {
+            // Update existingTask, in case we're not the first to acquire the lock
+            existingTask = Volatile.Read(ref _cacheableKeyRingTask);
+            if (existingTask is null)
+            {
+                // If there's no existing task, make one now
+                // PERF: Closing over utcNow substantially slows down the fast case (valid cache) in micro-benchmarks
+                // (closing over `this` for CacheableKeyRingProvider doesn't seem impactful)
+                existingTask = Task.Factory.StartNew(
+                    utcNow => CacheableKeyRingProvider.GetCacheableKeyRing((DateTime)utcNow!),
+                    utcNow,
+                    CancellationToken.None, // GetKeyRingFromCompletedTask will need to react if this becomes cancellable
+                    TaskCreationOptions.DenyChildAttach,
+                    TaskScheduler.Default);
+                Volatile.Write(ref _cacheableKeyRingTask, existingTask);
+            }
+        }
+
+        if (existingCacheableKeyRing is not null)
+        {
+            Debug.Assert(!forceRefresh, "Read cached key ring even though forceRefresh is true");
+            Debug.Assert(!CacheableKeyRing.IsValid(existingCacheableKeyRing, utcNow), "Should already have returned a valid cached key ring");
+            return existingCacheableKeyRing.KeyRing;
+        }
+
+        // Since there's no cached key ring we can return, we have to wait.  It's not ideal to wait for a task we
+        // just scheduled, but it makes the code a lot simpler (compared to having a separate, synchronous code path).
+        // Cleverness: swallow any exceptions - they'll be surfaced by GetKeyRingFromCompletedTask, if appropriate.
+        existingTask
+            .ContinueWith(static _ => { }, TaskScheduler.Default)
+            .Wait();
+
+        var newKeyRing = GetKeyRingFromCompletedTask(existingTask, utcNow); // Throws if the task failed (winning thread only)
+        if (newKeyRing is null)
+        {
+            // Another thread won - check whether it cached a new key ring
+            var newCacheableKeyRing = Volatile.Read(ref _cacheableKeyRing);
+            if (newCacheableKeyRing is null)
+            {
+                // There will have been a better exception from the winning thread
+                throw Error.KeyRingProvider_RefreshFailedOnOtherThread();
```

Comment:
The exception must have already happened at this point. Why can't we capture it and throw it from any thread that fails to retrieve the KeyRing due to the exception?

Reply:
I didn't think it would be helpful to print the underlying exception once per affected task, even if we were to add a note that it was being re-reported from elsewhere.

Reply:
I still think it's more helpful than throwing a generic exception.

Reply:
As a compromise, I've attached `existingTask.Exception` as the inner exception.